encoding/ssh/filexfer/buffer.go

```diff
@@ -0,0 +1,268 @@
+package filexfer
+
+import (
+	"encoding/binary"
+	"errors"
+)
+
+// Various encoding errors.
+var (
+	ErrShortPacket = errors.New("packet too short")
+	ErrLongPacket  = errors.New("packet too long")
+)
+
+// Buffer wraps up the various encoding details of the SSH format.
+//
+// Data types are encoded as per section 4 from https://tools.ietf.org/html/draft-ietf-secsh-architecture-09#page-8
+type Buffer struct {
+	b   []byte
+	off int
+}
+
+// NewBuffer creates and initializes a new Buffer using buf as its initial contents.
+func NewBuffer(b []byte) *Buffer {
+	return &Buffer{
+		b: b,
+	}
+}
+
+// NewMarshalBuffer creates an initializes a new Buffer ready to start marshaling a Packet into.
+// It prepopulates 4 bytes for length, the 1-byte packetType, and the 4-byte requestID.
+// It preallocates enough space for an additional size bytes of data above the prepopulated values.
+func NewMarshalBuffer(packetType PacketType, requestID uint32, size int) *Buffer {
+	buf := NewBuffer(make([]byte, 4, 4+1+4+size))
+
+	buf.AppendUint8(uint8(packetType))
+	buf.AppendUint32(requestID)
+
+	return buf
+}
+
+// Bytes returns a slice of length b.Len() holding the unconsumed bytes in the Buffer.
+func (b *Buffer) Bytes() []byte {
+	return b.b[b.off:]
+}
+
+// Packet finalizes the packet started from NewMarshalPacket.
+//
+// It writes the packet body length into the first four bytes of the Buffer in network byte order (big endian).
+// The packet body length is the size of the Buffer less the 4-byte length itself, plus the length of payload.
+func (b *Buffer) Packet(payload []byte) ([]byte, []byte, error) {
+	b.PutLength(len(b.b) - 4 + len(payload))
+
+	return b.b, payload, nil
+}
+
+// Len returns the number of unconsumed bytes in the Buffer.
+func (b *Buffer) Len() int {
+	return len(b.b) - b.off
+}
+
+// Reset resets the buffer to be empty, but it retains the underlying storage for use by future writes.
+func (b *Buffer) Reset() {
+	b.b = b.b[:0]
+	b.off = 0
+}
+
+// ConsumeUint8 consumes a single byte from the Buffer.
+// If Buffer does not have enough data, it will return ErrShortPacket.
+func (b *Buffer) ConsumeUint8() (uint8, error) {
+	if b.Len() < 1 {
+		return 0, ErrShortPacket
+	}
+
+	var v uint8
+	v, b.off = b.b[b.off], b.off+1
+	return v, nil
+}
+
+// AppendUint8 appends a single byte into the Buffer.
+func (b *Buffer) AppendUint8(v uint8) {
+	b.b = append(b.b, v)
+}
+
+// ConsumeBool consumes a single byte from the Buffer, and returns true if that byte is non-zero.
+// If Buffer does not have enough data, it will return ErrShortPacket.
+func (b *Buffer) ConsumeBool() (bool, error) {
```

Comment:
This method is never used. I don't think filexfer uses booleans at all (?).

Reply:
While `filexfer-02` does not use this type, it does appear in `filexfer-06`. While the protocol of this version of the library is at `filexfer-02`, if someone wanted to implement a later version of the protocol, I expect that most of these types and values will be directly imported in, rather than reimplementing everything. Thus, a few of these types have been written to support not-just-the-code-here, but any potential version of the `filexfer` format.